mono/utils/mono-os-wait-win32.c

```diff
@@ -0,0 +1,217 @@
+/**
+* \file
+* Win32 OS wait wrappers and interrupt/abort APC handling.
+*
+* Author:
+*   Johan Lorensson (lateralusx.github@gmail.com)
+*
+* Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+#include <mono/utils/mono-os-wait.h>
+#include <mono/utils/mono-threads.h>
+#include <mono/utils/mono-threads-debug.h>
+
+enum ThreadWaitInfo {
+	THREAD_WAIT_INFO_CLEARED = 0,
```

Comment:
I don't get why we need 3 APC queues, can't we get away with a single one?

Reply:
The first slot is used to indicate if a thread is currently in a OS wait call or not. Currently not used, but was put there for future use to detect if a thread is currently in an alertable OS wait call and potential optimize the APC queuing. The interrupt and abort slots was added since we had these two different APC types already in the code and used, so I didn't want to remove the possibility to differentiate them since that's what current code did, still making it possible to have both in flight at the same time.